app/services/wallet_transactions/mark_as_failed_service.rb

```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+module WalletTransactions
+  class MarkAsFailedService < BaseService
+    def initialize(wallet_transaction:)
+      @wallet_transaction = wallet_transaction
+      super()
+    end
+
+    def call
+      return result unless wallet_transaction
+      return result if wallet_transaction.status == "failed"
+
+      ActiveRecord::Base.transaction do
+        wallet_transaction.mark_as_failed!
+      end
```

Comment:
Since you are only updating the status, you can skip the transaction:
```suggestion
      wallet_transaction.mark_as_failed!
```